src/order/spam-protection/google-recaptcha.ts

```diff
@@ -1,23 +1,90 @@
 
+export interface RecaptchaResult {
+    error?: Error;
+    token?: string;
+}
+
 export default class GoogleRecaptcha {
+    private _event$?: Subject<RecaptchaResult>;
+    private _recaptcha?: ReCaptchaV2.ReCaptcha;
+
     constructor(
-        private googleRecaptchaScriptLoader: GoogleRecaptchaScriptLoader
+        private googleRecaptchaScriptLoader: GoogleRecaptchaScriptLoader,
+        private mutationObserverFactory: MutationObserverFactory
     ) {}
 
-    render(containerId: string, sitekey: string, callbacks: GoogleRecaptchaCallbacks): Promise<void> {
+    load(containerId: string, sitekey: string): Promise<void> {
+        const event$ = new Subject<RecaptchaResult>();
+        this._event$ = event$;
+
         return this.googleRecaptchaScriptLoader.load()
             .then(recaptcha => {
                 recaptcha.render(containerId, {
                     sitekey,
-                    ...callbacks,
+                    size: 'invisible',
+                    callback: () => {
+                        event$.next({
+                            token: recaptcha.getResponse(),
+                        });
+                    },
+                    'error-callback': () => {
+                        event$.next({
+                            error: new Error('Unable to complete spam verification because Recaptcha encountered an error.'),
+                        });
+                    },
                 });
+
+                this._recaptcha = recaptcha;
             });
     }
-}
 
-interface GoogleRecaptchaCallbacks {
-    callback(token: string): void;
-    'error-callback'?(): void;
-    'expired-callback'?(): void;
+    execute(): Observable<RecaptchaResult> {
+        if (!this._event$ || !this._recaptcha) {
+            throw new NotInitializedError(NotInitializedErrorType.SpamProtectionNotInitialized);
+        }
+
+        this._watchRecaptchaChallengeWindow();
+
+        this._recaptcha.execute();
+
+        return this._event$;
+    }
+
+    private _watchRecaptchaChallengeWindow() {
+        const iframeElement = document.querySelector('iframe[title="recaptcha challenge"]');
+
+        if (!iframeElement) {
+            throw new Error('Recaptcha challenge iframe not found.');
+        }
+
+        const iframeContainer = iframeElement.parentElement;
+
+        if (!iframeContainer) {
+            throw new Error('Recaptcha challenge iframe container not found.');
+        }
+
+        const container = iframeContainer.parentElement;
+
+        if (!container) {
+            throw new Error('Recaptcha challenge container not found.');
+        }
+
+        this.mutationObserverFactory.create(() => {
+            if (container.style.visibility === 'hidden') {
+                if (!this._event$) {
+                    throw new Error('Event is undefined.');
+                }
+
+                this._event$.next({
+                    error: new Error('Recaptcha challenge is not completed.'),
```

Comment:
might be good to throw a different type of error here, compared to the error you throw in `error-callback`. So SDK clients can decide how to handle these two different errors.

by checking the type, you should be able to present slightly different errors in the modal boxes.

> #### Spam Protection Failed
> We were not able to complete our spam protection verification. Please try again.

It would also be good to add a small comment explaining whats the use case you are trying to target with this piece of code as is not obvious.

Reply:
Great suggestion. Done.